File: apps/mobile/smarttaxi_app/lib/features/shared/stand_location.dart
import 'package:geolocator/geolocator.dart';
import 'models.dart';

/// Only an actual, recent, accurate device fix may join/renew a stand place.
/// Region centres and cached coordinates may sort the list but are not proof
/// of presence. Thirty seconds allows the parked-car 25-second heartbeat.
Coordinate? freshStandPosition(Position? position, DateTime now) {
  if (position == null ||
      !position.latitude.isFinite ||
      !position.longitude.isFinite ||
      position.latitude.abs() > 90 ||
      position.longitude.abs() > 180 ||
      !position.accuracy.isFinite ||
      position.accuracy < 0 ||
      position.accuracy > 60) {
    return null;
  }
  final age = now.difference(position.timestamp);
  if (age < const Duration(seconds: -5) || age > const Duration(seconds: 30)) {
    return null;
  }
  return Coordinate(lat: position.latitude, lng: position.longitude);
}

File: apps/web/src/features/shared/standLocation.mjs
// A parked car may report less often than a moving navigator. The stand's
// heartbeat is 25 seconds; only an actual sensor fix within 30 seconds counts.
export function freshStandPosition(position, now = Date.now()) {
  if (!position || !Number.isFinite(position.lat) || !Number.isFinite(position.lng) ||
      Math.abs(position.lat) > 90 || Math.abs(position.lng) > 180 ||
      !Number.isFinite(position.timestamp) || now - position.timestamp < -5000 ||
      now - position.timestamp > 30000 || !Number.isFinite(position.accuracy) ||
      position.accuracy < 0 || position.accuracy > 60) return null;
  return { lat: position.lat, lng: position.lng };
}
